server/src/main/java/org/opensearch/rest/NamedRoute.java

```diff
@@ -9,47 +9,168 @@
+        }
+
+        /**
+         * Sets the name for the route.
+         *
+         * @param name the name for the route
+         * @return the builder instance
+         */
+        public Builder uniqueName(String name) {
+            this.uniqueName = name;
+            return this;
+        }
+
+        /**
+         * Sets the legacy action names for the route.
+         *
+         * @param legacyActionNames the legacy action names for the route
+         * @return the builder instance
+         */
+        public Builder legacyActionNames(Set<String> legacyActionNames) {
+            this.legacyActionNames.addAll(validateLegacyActionNames(legacyActionNames));
+            return this;
+        }
+
+        /**
+         * Sets the handler for this route
+         *
+         * @param handler the handler for this route
+         * @return the builder instance
+         */
+        public Builder handler(Function<RestRequest, RestResponse> handler) {
+            if (handler == null) {
+                throw new OpenSearchException("Route handler must not be null.");
+            }
+            this.handler = handler;
+            return this;
         }
-        return routeName.matches(VALID_ACTION_NAME_PATTERN);
+
+        /**
+         * Builds a new instance of {@link NamedRoute} based on the provided parameters.
+         *
+         * @return a new instance of {@link NamedRoute}
+         * @throws OpenSearchException if the route name is invalid
+         */
+        public NamedRoute build() {
+            return new NamedRoute(this);
+        }
+
+        private Set<String> validateLegacyActionNames(Set<String> legacyActionNames) {
+            if (legacyActionNames == null) {
+                return new HashSet<>();
+            }
+            for (String actionName : legacyActionNames) {
+                if (!TransportService.isValidActionName(actionName)) {
+                    throw new OpenSearchException(
+                        "Invalid action name [" + actionName + "]. It must start with one of: " + TransportService.VALID_ACTION_PREFIXES
+                    );
+                }
+            }
+            return legacyActionNames;
+        }
+
     }
 
-    public NamedRoute(RestRequest.Method method, String path, String name) {
-        super(method, path);
-        if (!isValidRouteName(name)) {
+    private NamedRoute(Builder builder) {
+        super(builder.method, builder.path);
+        if (!isValidRouteName(builder.uniqueName)) {
             throw new OpenSearchException(
                 "Invalid route name specified. The route name may include the following characters"
                     + " 'a-z', 'A-Z', '0-9', ':', '/', '*', '_' and be less than "
                     + MAX_LENGTH_OF_ACTION_NAME
                     + " characters"
             );
         }
-        this.name = name;
+        this.uniqueName = builder.uniqueName;
+        this.actionNames = builder.legacyActionNames;
```

Comment:
As currently written, this set is still mutable and if you modify the set after building, it will change your object.  For example:
```java
builder = new NamedRoute.Builder();
builder.legacyActionNames(Set.of("foo"));
NamedRoute nr = builder.build();
builder.legacyActionNames(Set.of("bar"));
// nr.getActionNames() will contain "bar"
```
```suggestion
        this.actionNames = Set.copyOf(builder.legacyActionNames);
```